### PR TITLE
SLI-50 Panel to manage SonarQube server configurations

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: src/main/java/org/sonarlint/intellij/config/global/ConnectionTestTask.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+      Path tmp = Files.createTempDirectory("sonarlint-test");
+      GlobalConfiguration globalConfig = GlobalConfiguration.builder()
+        .setServerId("test")
+        .setVerbose(true)
+        .setWorkDir(tmp)
+        .build();
+
+      ServerConfiguration.Builder serverConfigBuilder = ServerConfiguration.builder()
+        .userAgent("SonarLint IntelliJ " + sonarlint.getVersion())
+        .connectTimeoutMilliseconds(2000)
```
Comment (henryju) [inline src/main/java/org/sonarlint/intellij/config/global/ConnectionTestTask.java]: Is it not too low?

Comment (dbmeneses) [inline src/main/java/org/sonarlint/intellij/config/global/ConnectionTestTask.java]: Agree, I will increase it. I got too inspired by other IntelliJ connection tests, which use 2s only.

**Comment by henryju** [comment]
LGTM

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerEditorPanel.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    server.setServerId(idText.getText().trim());
+    server.setHostUrl(urlText.getText().trim());
+    server.setLogin(loginText.getText().trim());
+    server.setPassword(new String(passwordText.getPassword()));
+    server.setEnableProxy(enableProxy.isSelected());
+
+    changeListener.consume(server);
+  }
+
+  @Override public void dispose() {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerEditorPanel.java]: MAJOR Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1186)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerEditorPanel.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+import com.intellij.ui.components.JBTextField;
+import com.intellij.util.Consumer;
+import com.intellij.util.net.HttpConfigurable;
+import com.intellij.util.ui.FormBuilder;
+
+import java.awt.Component;
+import java.awt.GridBagConstraints;
+import java.awt.GridBagLayout;
+import java.awt.Insets;
+import javax.swing.*;
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerEditorPanel.java]: MAJOR Explicitly import the specific classes needed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2208)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerEditorPanel.java [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+import com.intellij.ui.components.JBPasswordField;
+import com.intellij.ui.components.JBTextField;
+import com.intellij.util.Consumer;
+import com.intellij.util.net.HttpConfigurable;
+import com.intellij.util.ui.FormBuilder;
+
+import java.awt.Component;
+import java.awt.GridBagConstraints;
+import java.awt.GridBagLayout;
+import java.awt.Insets;
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerEditorPanel.java]: MINOR Remove this unused import 'java.awt.Insets'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerEditorPanel.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+      public void actionPerformed(ActionEvent e) {
+        testConnection();
+      }
+    });
+
+    final JPanel p = new JPanel(new GridBagLayout());
+    addForm(p);
+
+    proxySettingsButton.addActionListener(new ActionListener() {
+      public void actionPerformed(ActionEvent e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerEditorPanel.java]: MAJOR Add the "@Override" annotation above this method signature [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1161)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerEditorPanel.java [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    enableProxy = new JBCheckBox("Use proxy", server.enableProxy());
+    enableProxy.setMnemonic('y');
+
+    enableProxy.setEnabled(HttpConfigurable.getInstance().USE_HTTP_PROXY);
+
+    testButton = new JButton("Test connection");
+    testButton.setHorizontalAlignment(SwingConstants.RIGHT);
+    testButton.setAlignmentX(Component.RIGHT_ALIGNMENT);
+    testButton.addActionListener(new ActionListener() {
+      public void actionPerformed(ActionEvent e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerEditorPanel.java]: MAJOR Add the "@Override" annotation above this method signature [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1161)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerEditorPanel.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    changeListener.consume(server);
+  }
+
+  @Override public void dispose() {
+  }
+
+  /**
+   * A {@link FormBuilder} that doesn't fill buttons
+   */
+  private class CustomFormBuilder extends FormBuilder {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerEditorPanel.java]: MAJOR Make this a "static" inner class. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2694)

File: src/main/java/org/sonarlint/intellij/config/global/SonarLintGlobalSettings.java [hunk 8]
```diff
@@ -21,21 +21,21 @@
 
 import com.intellij.openapi.application.ApplicationManager;
 import com.intellij.openapi.application.PathManager;
-import com.intellij.openapi.components.ExportableApplicationComponent;
-import com.intellij.openapi.components.PersistentStateComponent;
-import com.intellij.openapi.components.State;
-import com.intellij.openapi.components.Storage;
-import com.intellij.openapi.components.StoragePathMacros;
+import com.intellij.openapi.components.*;
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarLintGlobalSettings.java]: MAJOR Explicitly import the specific classes needed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2208)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServer.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    setPassword(toCopy.getPassword());
+    setToken(toCopy.getToken());
+    setLogin(toCopy.getLogin());
+    setName(toCopy.getName());
+    setEnableProxy(toCopy.enableProxy());
+  }
+
+  @Override
+  public boolean equals(Object o){
+    if (!(o instanceof SonarQubeServer)) return false;
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServer.java]: MAJOR At most one statement is allowed per line, but 2 statements were found on this line. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00122)
MINOR Missing curly brace. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00121)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServer.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    setServerId(toCopy.getServerId());
+    setPassword(toCopy.getPassword());
+    setToken(toCopy.getToken());
+    setLogin(toCopy.getLogin());
+    setName(toCopy.getName());
+    setEnableProxy(toCopy.enableProxy());
+  }
+
+  @Override
+  public boolean equals(Object o){
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServer.java]: CRITICAL This class overrides "equals()" and should therefore also override "hashCode()". [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1206)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServer.java [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+  private String hostUrl;
+  private String token;
+
+  private String name;
+  private String login;
+  private String password;
+  private String serverId;
+  private boolean enableProxy;
+
+  public SonarQubeServer() {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServer.java]: MAJOR Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1186)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServer.java [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+  @Tag("token")
+  public String getEncodedToken() {
+    return PasswordUtil.encodePassword(getToken());
+  }
+
+  public void setEncodedToken(String token) {
+    try {
+      setToken(PasswordUtil.decodePassword(token));
+    }
+    catch (NumberFormatException e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServer.java]: MINOR Move this "catch" on the same line that the previous closing curly brace. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ARightCurlyBraceSameLineAsNextBlockCheck)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServer.java [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+    setName(toCopy.getName());
+    setEnableProxy(toCopy.enableProxy());
+  }
+
+  @Override
+  public boolean equals(Object o){
+    if (!(o instanceof SonarQubeServer)) return false;
+    SonarQubeServer other = (SonarQubeServer)o;
+
+    return Comparing.equal(getHostUrl(), other.getHostUrl()) &&
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServer.java]: MAJOR Reduce the number of conditional operators (6) used in the expression (maximum allowed 3). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1067)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServer.java [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+  @Tag("password")
+  public String getEncodedPassword() {
+    return PasswordUtil.encodePassword(getPassword());
+  }
+
+  public void setEncodedPassword(String password) {
+    try {
+      setPassword(PasswordUtil.decodePassword(password));
+    }
+    catch (NumberFormatException e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServer.java]: MINOR Move this "catch" on the same line that the previous closing curly brace. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ARightCurlyBraceSameLineAsNextBlockCheck)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+        DefaultActionGroup group = new DefaultActionGroup();
+        group.add(new AddServerAction());
+
+        JBPopupFactory.getInstance()
+          .createActionGroupPopup(LABEL_ADD_SERVER, group, DataManager.getInstance().getDataContext(anActionButton.getContextComponent()),
+            JBPopupFactory.ActionSelectionAid.SPEEDSEARCH, true).show(anActionButton.getPreferredPopupPoint());
+      }
+    });
+
+    toolbarDecorator.setRemoveAction(new AnActionButtonRunnable() {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java]: MAJOR Reduce this anonymous class number of lines from 21 to at most 20, or make it a named class. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1188)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+
+  // Model
+  private List<SonarQubeServer> servers = new ArrayList<>();
+  private Consumer<SonarQubeServer> changeListener;
+
+  private final FactoryMap<SonarQubeServer, String> serverIds = new ConcurrentFactoryMap<SonarQubeServer, String>() {
+    private int count;
+    @Override
+    protected String create(SonarQubeServer server) {
+      return Integer.toString(count++);
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java]: MAJOR Extract this increment or decrement operator into a dedicated statement. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS881)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+
+    JBLabel emptyLabel = new JBLabel("No server selected", SwingConstants.CENTER);
+    emptyPanel = new JPanel(new BorderLayout());
+    emptyPanel.add(emptyLabel, BorderLayout.CENTER);
+
+    panel = new JPanel(new BorderLayout());
+    panel.add(splitter);
+
+    serverList.getSelectionModel().addListSelectionListener(new ListSelectionListener() {
+      public void valueChanged(@NotNull ListSelectionEvent e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java]: MAJOR Add the "@Override" annotation above this method signature [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1161)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+          url = LABEL_UNDEFINED;
+        } else {
+          url = server.getName();
+        }
+        append(url, SimpleTextAttributes.REGULAR_ATTRIBUTES);
+      }
+    });
+
+    changeListener = new Consumer<SonarQubeServer>() {
+      public void consume(SonarQubeServer server) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java]: MAJOR Add the "@Override" annotation above this method signature [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1161)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+import com.intellij.util.containers.FactoryMap;
+import java.util.LinkedList;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.sonarlint.intellij.util.ResourceLoader;
+
+import javax.swing.*;
+import javax.swing.event.ListSelectionEvent;
+import javax.swing.event.ListSelectionListener;
+import java.awt.*;
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java]: MAJOR Explicitly import the specific classes needed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2208)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java [hunk 20]
```diff
@@ -0,0 +1,258 @@
...
+  private void addServerEditor(SonarQubeServer server) {
+    SonarQubeServerEditorPanel editor = new SonarQubeServerEditorPanel(changeListener, server);
+
+    JComponent component = editor.create();
+    String name = serverIds.get(server);
+    editorPanel.add(component, name);
+    editorPanel.doLayout();
+  }
+
+  private void addServer(SonarQubeServer server) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java]: MAJOR Move this method into "AddServerAction". [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS3398)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+import com.intellij.ui.components.JBList;
+import com.intellij.util.Consumer;
+import com.intellij.util.containers.ConcurrentFactoryMap;
+import com.intellij.util.containers.FactoryMap;
+import java.util.LinkedList;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.sonarlint.intellij.util.ResourceLoader;
+
+import javax.swing.*;
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java]: MAJOR Explicitly import the specific classes needed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2208)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+
+        CollectionListModel model = (CollectionListModel) serverList.getModel();
+        // it's not removed from serverIds and editorList
+        model.remove(server);
+        servers.remove(server);
+
+        if (model.getSize() > 0) {
+          serverList.setSelectedValue(model.getElementAt(0), true);
+        }
+        else {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java]: MINOR Move this "else" on the same line that the previous closing curly brace. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ARightCurlyBraceSameLineAsNextBlockCheck)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+      return Integer.toString(count++);
+    }
+  };
+
+  public SonarQubeServerMgmtPanel(SonarLintGlobalSettings settings) {
+    create();
+    load(settings);
+  }
+
+  public void create() {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java]: MAJOR The Cyclomatic Complexity of this method "create" is 13 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java [hunk 24]
```diff
@@ -0,0 +1,258 @@
...
+      }
+    });
+
+    serverList.setCellRenderer(new ColoredListCellRenderer() {
+      @Override
+      protected void customizeCellRenderer(JList list, Object value, int index, boolean selected, boolean hasFocus) {
+        SonarQubeServer server = (SonarQubeServer)value;
+        try {
+          setIcon(ResourceLoader.getIcon(LIST_ICON));
+        } catch (IOException e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java]: CRITICAL Either log or rethrow this exception. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java [hunk 25]
```diff
@@ -0,0 +1,258 @@
...
+      super("SonarQube server", "New SonarQube server", null);
+      setIcon();
+    }
+
+    private void setIcon() {
+      try {
+        Icon icon = ResourceLoader.getIcon(LIST_ICON);
+        super.getTemplatePresentation().setIcon(icon);
+      } catch (IOException e) {
+        e.printStackTrace();
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java]: CRITICAL Use a logger to log this exception. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java [hunk 26]
```diff
@@ -0,0 +1,258 @@
...
+    });
+
+    serverList.setCellRenderer(new ColoredListCellRenderer() {
+      @Override
+      protected void customizeCellRenderer(JList list, Object value, int index, boolean selected, boolean hasFocus) {
+        SonarQubeServer server = (SonarQubeServer)value;
+        try {
+          setIcon(ResourceLoader.getIcon(LIST_ICON));
+        } catch (IOException e) {
+          e.printStackTrace();
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java]: CRITICAL Use a logger to log this exception. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java [hunk 27]
```diff
@@ -0,0 +1,258 @@
...
+    public AddServerAction() {
+      super("SonarQube server", "New SonarQube server", null);
+      setIcon();
+    }
+
+    private void setIcon() {
+      try {
+        Icon icon = ResourceLoader.getIcon(LIST_ICON);
+        super.getTemplatePresentation().setIcon(icon);
+      } catch (IOException e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java]: CRITICAL Either log or rethrow this exception. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java [hunk 28]
```diff
@@ -0,0 +1,258 @@
...
+package org.sonarlint.intellij.config.global;
+
+import com.intellij.ide.DataManager;
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.actionSystem.DefaultActionGroup;
+import com.intellij.openapi.project.DumbAware;
+import com.intellij.openapi.roots.ui.configuration.actions.IconWithTextAction;
+import com.intellij.openapi.ui.Splitter;
+import com.intellij.openapi.ui.popup.JBPopupFactory;
+import com.intellij.ui.*;
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/config/global/SonarQubeServerMgmtPanel.java]: MAJOR Explicitly import the specific classes needed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2208)

**Comment by SonarTech** [comment]
SonarQube analysis reported 28 issues:
* CRITICAL 5 critical
* MAJOR 18 major
* MINOR 5 minor

Watch the comments in this conversation to review them.

**Pull request merged**